L46-core-tls-credential-API.md

```diff
@@ -0,0 +1,463 @@
+  // sync_status: This will only be useful if the verifier is synchronous.
+  //              The status of the verifier as it has already done it's
+  //              synchronous check.
+  // return: return true if executed synchronously, otherwise return false
+  bool Verify(TlsCustomVerificationCheckRequest* request,
+              std::function<void(Status)> callback,
+              Status* sync_status);
+
+  // Cancels a verification request previously started via Verify().
+  // Used when the connection attempt times out or is cancelled while an async
+  // verification request is pending.
+  //
+  // request: the verification information associated with this request
+  void Cancel(TlsCustomVerificationCheckRequest* request);
+};
+
+// A CertificateVerifier that doesn't perform any additional checks other than
+// certificate verification, if specified.
+// Note: using this solely without any other authentication mechanisms on the
+// peer identity will leave your applications to the MITM(Man-In-The-Middle)
+// attacks. Users should avoid doing so in production environments.
+class NoOpCertificateVerifier : public CertificateVerifier {
+ public:
+  NoOpCertificateVerifier();
+};
+
+// A CertificateVerifier that will perform hostname verification, to see if the
+// target name set from the client side matches the identity information
+// specified on the server's certificate.
+class HostNameCertificateVerifier : public CertificateVerifier {
+ public:
+  HostNameCertificateVerifier();
+};
+```
+
+#### Custom Chain Building
+```c++
+class CustomChainBuilderInterface {
+public:
+ // TODO(gtcooke94) - more detail here on output specifics.
+ // The output chain should consist of DER encoded certs ordered in the same way
+ // that Open/BoringSSL build chains in `X509_verify_cert` including the root of
+ // the trusted chain.
+ virtual absl::StatusOr<std::vector<absl::string_view>>
+ BuildAndVerifyChain(const std::vector<absl::string_view>& peer_cert_chain_der)
+ = 0;
+ virtual ~CustomChainBuilderInterface() = 0;
+}
+```
+
+The default behavior for chain building is based on the underlying SSL library. For example, [X509_verify_cert](https://www.openssl.org/docs/man1.1.1/man3/X509_verify_cert.html) is the implementation in OpenSSL that builds and verifies a certificate chain.
+gRPC calls that function by default [here](https://github.com/grpc/grpc/blob/2d2d5a3c411a2bade319a08085e55821cf2d5ed9/src/core/tsi/ssl_transport_security.cc#L1150).
+Once the `CustomChainBuilder` is implemented, this will be where it is used. The `X509_verify_cert` call will be replaced by the `CustomChainBuilder's BuildAndVerifyChain` that fully replaces chain building.
+Replacing chain building is very complex and can open dangerous security holes if done wrong. Most users should not expect to use this API. However, there exist use cases for such behavior - for example, building chains from SPIFFE IDs and trust maps.
```

Comment:
nit: These last couple of sentences are replacing the intro from the section, so they could be omitted.